pallets/balances/src/lib.rs

```diff
@@ -462,6 +462,22 @@ decl_module! {
             let dest = T::Lookup::lookup(dest)?;
             Self::transfer_core(&source, &dest, value, None, ExistenceRequirement::AllowDeath)?;
         }
+
+        /// Burns the given amount of tokens from the caller's free balance.
+        #[weight = SimpleDispatchInfo::FixedNormal(200_000)]
+        pub fn burn_free_balance(origin, amount: T::Balance) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let total_issuance = Self::total_issuance();
+            let amount = cmp::min(total_issuance, cmp::min(amount, Self::account(&who).free));
+            // Burn total issuance.
+            <TotalIssuance<T>>::put(total_issuance - amount);
+            // Update the caller's free balance.
```

Comment:
`ensure_can_withdraw` check should be done before updating balance to make sure enough "unlocked" balance is available. `free` balance is sum of both locked and unlocked balances.

Reply:
Same problem here: there is no bit flag in `WithdrawReasons` for burning. I'll try the 0 bitmask.

Reply:
I think we should use the `transfer` reason. If a user is allowed to transfer, they should be allowed to burn.